lib/datadog/statsd/single_thread_sender.rb

```diff
@@ -2,12 +2,24 @@
 
 module Datadog
   class Statsd
+    # The SingleThreadSender is a sender synchronously buffering messages
+    # in a `MessageBuffer`.
+    # It is using current Process.PID to check it is the result of a recent fork
+    # and it is reseting the MessageBuffer if that's the case.
     class SingleThreadSender
       def initialize(message_buffer)
         @message_buffer = message_buffer
+        # store the pid for which this sender has been created
+        update_fork_pid
       end
 
       def add(message)
+        # we have just forked, meaning we have messages in the buffer that we should
+        # not send, they belong to the parent process, let's clear the buffer.
+        if forked?
+          @message_buffer.reset
+          update_fork_pid
+        end
```

Comment:
A similar race to the one I described for the `Sender` can happen here. Two threads call this method at the same time, both of them get `forked?` => `true` and the `@message_buffer` gets reset twice.

Also the interaction with the buffer itself may have issues with concurrency.